src/NonCumulative/Soundness/Properties/NoFunExt/LogRel.agda

```diff
@@ -0,0 +1,161 @@
+open import NonCumulative.Soundness.LogRel
+open import NonCumulative.Statics.Ascribed.Presup
+open import NonCumulative.Statics.Ascribed.Misc
+open import NonCumulative.Statics.Ascribed.Refl
+open import NonCumulative.Statics.Ascribed.CtxEquiv
+open import NonCumulative.Statics.Ascribed.Properties
+open import NonCumulative.Statics.Ascribed.Properties.Contexts
+
+-----------------------------------------------
+-- Properties of the gluing model for natural numbers
+
+®Nat⇒∈Nat : Γ ⊢ t ∶N® a ∈Nat → a ∈′ Nat
+®Nat⇒∈Nat (ze t≈)    = ze
+®Nat⇒∈Nat (su _ rel) = su (®Nat⇒∈Nat rel)
+®Nat⇒∈Nat (ne c∈ _)  = ne c∈
+
+®Nat⇒∶Nat : Γ ⊢ t ∶N® a ∈Nat → ⊢ Γ → Γ ⊢ t ∶[ 0 ] N
+®Nat⇒∶Nat (ze t≈) ⊢Γ    = proj₁ (proj₂ (presup-≈ t≈))
+®Nat⇒∶Nat (su t≈ _) ⊢Γ  = proj₁ (proj₂ (presup-≈ t≈))
+®Nat⇒∶Nat (ne _ rel) ⊢Γ = [I]-inv (proj₁ (proj₂ (presup-≈ (rel (⊢wI ⊢Γ)))))
+
+®Nat-resp-≈ : Γ ⊢ t ∶N® a ∈Nat → Γ ⊢ t ≈ t′ ∶[ 0 ] N →  Γ ⊢ t′ ∶N® a ∈Nat
+®Nat-resp-≈ (ze t≈) t≈t′     = ze (≈-trans (≈-sym t≈t′) t≈)
+®Nat-resp-≈ (su t≈ t∼a) t≈t′ = su (≈-trans (≈-sym t≈t′) t≈) t∼a
+®Nat-resp-≈ (ne c∈ rel) t≈t′ = ne c∈ λ ⊢σ → ≈-trans ([]-cong-N′ (≈-sym t≈t′) (⊢w⇒⊢s ⊢σ)) (rel ⊢σ)
+
+®Nat-resp-⊢≈ : Γ ⊢ t ∶N® a ∈Nat → ⊢ Γ ≈ Δ →  Δ ⊢ t ∶N® a ∈Nat
+®Nat-resp-⊢≈ (ze t≈) Γ≈Δ     = ze (ctxeq-≈ Γ≈Δ t≈)
+®Nat-resp-⊢≈ (su t≈ t∼a) Γ≈Δ = su (ctxeq-≈ Γ≈Δ t≈) (®Nat-resp-⊢≈ t∼a Γ≈Δ)
+®Nat-resp-⊢≈ (ne c∈ rel) Γ≈Δ = ne c∈ (λ ⊢σ → rel (⊢w-resp-⊢≈ʳ ⊢σ (⊢≈-sym Γ≈Δ)))
+
+-- we prove this lemma directly so we do not have to rely on realizability of the PER
+-- model which in turn relies on functional extensionality.
+®Nat⇒∈Top : Γ ⊢ t ∶N® a ∈Nat → ↓ 0 N a ∈′ Top
+®Nat⇒∈Top (ze t≈) ns     = ze , Rze ns , Rze ns
+®Nat⇒∈Top (su t≈ t′∼a) ns
+  with ®Nat⇒∈Top t′∼a ns
+...  | w , ↘w , ↘w′        = su w , Rsu ns ↘w , Rsu ns ↘w′
+®Nat⇒∈Top (ne c∈ rel) ns
+  with c∈ ns
+...  | u , ↘u , ↘u′ = ne u , RN ns ↘u′ , RN ns ↘u′
+
+-- If t and a are related as natural numbers, then t and the readback of a are equivalent up to any weakening.
+®Nat⇒≈ : (t∼a : Γ ⊢ t ∶N® a ∈Nat) → Δ ⊢w σ ∶ Γ → Δ ⊢ t [ σ ] ≈ Nf⇒Exp (proj₁ (®Nat⇒∈Top t∼a (len Δ))) ∶[ 0 ] N
+®Nat⇒≈ (ze t≈) ⊢σ     = ≈-trans ([]-cong-N′ t≈ (⊢w⇒⊢s ⊢σ)) (ze-[] (⊢w⇒⊢s ⊢σ))
+®Nat⇒≈ (su t≈ t′∼a) ⊢σ
+  with presup-s (⊢w⇒⊢s ⊢σ)
+...  | _ , ⊢Γ         = ≈-trans ([]-cong-N′ t≈ (⊢w⇒⊢s ⊢σ)) (≈-trans (su-[] (⊢w⇒⊢s ⊢σ) (®Nat⇒∶Nat t′∼a ⊢Γ)) (su-cong (®Nat⇒≈ t′∼a ⊢σ)))
+®Nat⇒≈ (ne c∈ rel) ⊢σ = rel ⊢σ
+
+----------------------------------
+-- Properties of the gluing models
+
+-- If T and A (and B) are related in level i, then T is typed in level i.
+®⇒ty : ∀ {i} (A≈B : A ≈ B ∈ 𝕌 i) →
+       Γ ⊢ T ®[ i ] A≈B →
+       -----------------------
+       Γ ⊢ T ∶[ 1 + i ] Se i
+®⇒ty (ne C≈C′ j≡1+i j′≡1+i) (⊢T , _) = ⊢T
+®⇒ty (N i≡0) T® = proj₁ (proj₂ (presup-≈ T®))
+®⇒ty (U i≡1+j j≡j′) T® = proj₁ (proj₂ (presup-≈ T®))
+®⇒ty (Π i≡maxjk jA RT j≡j′ k≡k′) record { IT = IT ; OT = OT ; ⊢OT = ⊢OT ; T≈ = T≈ ; krip = krip } = proj₁ (proj₂ (presup-≈ T≈))
```

Comment:
open a record in a where clause to avoid clutters. check other files for examples.

Reply:
see below, in fact